File: htdocs/request/grx/time_mot_loc.php
<?php
/*
 * Generate a GR placefile with TIME..MOT..LOC values 
 * $Id: $:
 */
include("../../../config/settings.inc.php");
include_once "../../../include/database.inc.php";
$plimit = isset($_GET["all"]) ? "('TO','SV')" : "('TO')";
$title = isset($_GET["all"]) ? "SVR+TOR" : "TOR";
$pgconn = iemdb('postgis');
$rs = pg_query($pgconn, "SELECT ST_x(tml_geom) as lon, ST_y(tml_geom) as lat,
	ST_AsText(tml_geom_line) as line,
	tml_valid at time zone 'UTC' as tmlv, tml_direction, tml_sknt, 
	polygon_end at time zone 'UTC' as pe, eventid, wfo, phenomena
	 from sbw_". date("Y") ." 
	WHERE polygon_end > now() and phenomena in $plimit and status != 'CAN'");

header( 'Content-type: text/plain');


putenv("TZ=GMT");

echo "RefreshSeconds: 60
Threshold: 999
Title: NWS $title Warning Time-Mot-Loc 
Font: 1, 11, 1, \"Courier New\" 

";

for ($k=0;$row=pg_fetch_array($rs);$k++){
	$valid = strtotime($row["tmlv"]);
	$pe = strtotime($row["pe"]);
	$dur = $pe - $valid;
	$min = $dur/60;
	$spd = floatval($row["tml_sknt"]) * 0.514444444;
	$dir = floatval($row["tml_direction"]);
	$descript = sprintf("%s/%s", $dir, $row["tml_sknt"]);
	$distance = $spd * $dur;

	/* Some magic if we have a line or a string! */
	$lats = Array();
	$lons = Array();
	if ($row["lat"]){
		$lats[] = floatval($row["lat"]);
		$lons[] = floatval($row["lon"]);
	} else if ($row["line"]){
		preg_match_all('/(?P<lon>[0-9\.\-]+) (?P<lat>[0-9\.\-]+)/', 
					$row["line"], $matches);
		$lats = $matches["lat"];
		$lons = $matches["lon"];
		echo "Color: 255 255 0 100\n";
		echo "Polygon:\n";
		$two = "";
		for($m=sizeof($lats)-1; $m > -1; $m--){
			echo sprintf("%s,%s\n", $lats[$m], $lons[$m]);
	  		$lon2 = round($lons[$m] + (($distance * cos(deg2rad(270-$dir)))/(111325 * cos(deg2rad($lats[$m])))),6);
	  		$lat2 = round($lats[$m] + (($distance * sin(deg2rad(270-$dir)))/111325),6);
			$two = sprintf("%s,%s\n%s", $lat2, $lon2, $two);
		}
		echo $two;
		echo "End:\n";
		
	}
	reset($lats);
	for($m=0; $m < sizeof($lats); $m++){
		$lat1 = $lats[$m];
		$lon1 = $lons[$m];
		if ($row["phenomena"] == 'SV'){
			echo "Color: 255 255 0\n";
		} else {
			echo "Color: 255 0 0\n";
		}
	  $lon2 = round($lon1 + (($distance * cos(deg2rad(270-$dir)))/(111325 * cos(deg2rad($lat1)))),6);
	  $lat2 = round($lat1 + (($distance * sin(deg2rad(270-$dir)))/111325),6);

	  echo "Line: 2,0,\"NWS Warning Track (". $row["wfo"] ."-". $row["phenomena"] 
	."-". $row["eventid"].")\\n".$descript."\"
  ".$lat1.",".$lon1."
  ".$lat2.",".$lon2."
End:

";
	  echo "Color: 255 255 255\nThreshold:10\n\n";

	for($i=0;$i<=$min;$i++){
		$now = $valid + ($i * 60);
		$dur = $now - $valid;
		$distance = $spd * $dur;
		$lon2 = round($lon1 + (($distance * cos(deg2rad(270-$dir)))/(111325 * cos(deg2rad($lat1)))),6);
		$lat2 = round($lat1 + (($distance * sin(deg2rad(270-$dir)))/111325),6);
		$t_now = date("Hi\z",$now);
	
		//echo "Object: ".$lat2.",".$lon2."\n  Text: 0,0,1,\"".$t_now."\"\nEnd:\n\n";
		echo "Place: ".$lat2.",".$lon2.",".$t_now."\n";
	}
	
	echo "Threshold: 999\n\n";
	}
}
?>

File: htdocs/request/grx/time_mot_loc.py
"""Serve Time Motion Location Vectors."""
import datetime
import math
import re
from io import StringIO
from zoneinfo import ZoneInfo

from pyiem.util import get_dbconnc, utc
from pyiem.webutil import iemapp

LOLA = re.compile(r"(?P<lon>[0-9\.\-]+) (?P<lat>[0-9\.\-]+)")
SQLISO = "YYYY-MM-DDThh24:MI:SSZ"


def extrapolate(lon, lat, distance, drct):
    """Fun."""
    lon2 = lon + (
        (distance * math.cos(math.radians(270.0 - drct)))
        / (111325.0 * math.cos(math.radians(lat)))
    )
    lat2 = lat + ((distance * math.sin(math.radians(270.0 - drct))) / 111325.0)
    return lon2, lat2


def gentext(sio, row, grversion):
    """Turn the database row into placefile worthy text."""
    smps = row["tml_sknt"] * 0.5144
    drct = row["tml_direction"]
    tml_valid = row["tml_valid"].astimezone(ZoneInfo("UTC"))
    duration = (row["polygon_end"] - row["tml_valid"]).total_seconds()
    distance = smps * duration
    lats = []
    lons = []
    time_range = ""
    if grversion >= 1.5:
        time_range = f"TimeRange: {row['iso_begin']} {row['iso_end']}\n"
    if row["lat"] is not None:
        lons.append(row["lon"])
        lats.append(row["lat"])
        # Point events are represented by lines
        lon2, lat2 = extrapolate(row["lon"], row["lat"], distance, drct)
        sio.write(
            f"Color: 255 255 0\n{time_range}"
            f"Line: 2,0,\"NWS Warning Track ({row['wfo']}-{row['phenomena']}"
            f"-{row['eventid']})\\n"
            f"{drct:.0f}/{row['tml_sknt']:.0f}\"\n"
            f"{row['lat']:.4f},{row['lon']:.4f}\n"
            f"{lat2:.4f},{lon2:.4f}\n"
            "End:\n"
        )
    elif row["line"] is not None:
        # Lines are shown as three sets of lines.
        for token in LOLA.findall(row["line"]):
            lons.append(float(token[0]))
            lats.append(float(token[1]))
        for seconds in [0, duration / 2.0, duration]:
            valid = tml_valid + datetime.timedelta(seconds=seconds)
            ts = valid.strftime("%H%Mz")
            sio.write(
                f"Color: 255 255 0\n{time_range}"
                f'Line: 2,0,"NWS Warning Track @{ts}'
                f"({row['wfo']}-{row['phenomena']}"
                f"-{row['eventid']})\\n"
                f"{drct:.0f}/{row['tml_sknt']:.0f}\"\n"
            )
            for lon, lat in zip(lons, lats):
                lon2, lat2 = extrapolate(lon, lat, seconds * smps, drct)
                sio.write(f"{lat2:.4f},{lon2:.4f}\n")
            sio.write("End:\n")

    # Now we place places along the track.
    sio.write("Color: 255 255 255\nThreshold:10\n\n")
    for lon, lat in zip(lons, lats):
        for minute in range(0, int(duration / 60.0) + 1):
            valid = tml_valid + datetime.timedelta(minutes=minute)
            ts = valid.strftime("%H%Mz")
            lon2, lat2 = extrapolate(lon, lat, smps * minute * 60, drct)
            sio.write(f"Place: {lat2:.4f},{lon2:.4f},{ts}\n")
    sio.write("Threshold: 999\n")


@iemapp()
def application(environ, start_response):
    """Our WSGI service."""
    grversion = float(environ.get("version", 1.0))
    is_all = "all" in environ
    phenoms = ["TO", "SV"] if is_all else ["TO"]
    pgconn, cursor = get_dbconnc("postgis")
    valid = utc()
    refresh = 60
    if "valid" in environ:
        # pylint: disable=no-value-for-parameter
        valid = datetime.datetime.strptime(
            environ.get("valid")[:16], "%Y-%m-%dT%H:%M"
        ).replace(tzinfo=ZoneInfo("UTC"))
        refresh = 86400
    t1 = valid
    t2 = valid
    tmlabel = valid.strftime("%H%Mz")
    if grversion >= 1.5 or "valid" in environ:
        # Pull larger window of data to support TimeRange
        t1 = valid - datetime.timedelta(hours=2)
        t2 = valid + datetime.timedelta(hours=2)
        tmlabel = valid.strftime("%b %d %Y %H%Mz")
    cursor.execute(
        f"""SELECT ST_x(tml_geom) as lon, ST_y(tml_geom) as lat,
        ST_AsText(tml_geom_line) as line,
        tml_valid, tml_direction, tml_sknt,
        polygon_begin, polygon_end, eventid, wfo, phenomena,
        to_char(polygon_begin at time zone 'UTC', '{SQLISO}') as iso_begin,
        to_char(polygon_end at time zone 'UTC', '{SQLISO}') as iso_end
        from sbw_{valid.year} WHERE polygon_end > %s and
        polygon_begin < %s and phenomena = ANY(%s)
        and tml_direction is not null and status != 'CAN' and
        polygon_end > polygon_begin""",
        (t1, t2, phenoms),
    )
    sio = StringIO()
    label = "TOR+SVR" if is_all else "TOR"
    sio.write(
        f"RefreshSeconds: {refresh}\n"
        "Threshold: 999\n"
        f"Title: NWS {label} @{tmlabel} Warning Time-Mot-Loc\n"
        'Font: 1, 11, 1, "Courier New"\n\n'
    )
    for row in cursor:
        gentext(sio, row, grversion)
    pgconn.close()
    start_response("200 OK", [("Content-type", "text/plain")])
    return [sio.getvalue().encode("utf-8")]
